[russoundserial] Russound Serial Binding initial contribution

holmes opened the pull request:
This binding allows you to control a Russound MCA/CAA system over a
serial connection. There is an existing Russound plugin, but that
works over a network connection on a different protocol (RIO).

This plugin uses a kotlin library to communicate via the serial protocol.
It's available at https://github.com/holmes/russound

Signed-off-by: Jason Holmes <holmes.j@gmail.com> (github: holmes)

holmes commented:
I'm not sure what's up w/ the signed-off-by. The line is definitely in my commit!

martinvw commented:
Maybe the email address does not match the email address you used when committing? I had this problem once too.

holmes commented:
@martinvw good call. it's using my work author tag instead of my personal one.

martinvw commented:
Could you please rebase on the current master there was a compilation error in the master 😊

holmes commented:
rebased nad pushed.

holmes commented:
@martinvw or @kaikreuzer would y'all take a look and approve?

triller-telekom commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java:
```diff
@@ -0,0 +1,89 @@
+
+    if (thingTypeUID.equals(SERIAL_BRIDGE)) {
+      String serialPort = (String) thing.getConfiguration().get("serialPort");
+      File deviceLocation = new File(serialPort);
```
This only works on Unix machines, right? On windows we cannot talk to the serial port via a file handle, can we? Or do you handle this somehow in your `SerialCommandSender`? If its a limitation I would briefly mention it in the README file.

triller-telekom commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java:
```diff
@@ -0,0 +1,89 @@
+        logger.info("Couldn't find serialPort at {}", serialPort, e);
+        return null;
+      } catch (Exception e) {
+        logger.info("Unable to create RussoundSerialBridgeHandler", e);
```
A user reading such an error message can probably not do anything about it, so please reduce the severity to `debug`.

triller-telekom commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java:
```diff
@@ -0,0 +1,89 @@
+ * @author holmes - Initial contribution
+ */
+public class RussoundSerialHandlerFactory extends BaseThingHandlerFactory {
+  private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS =
```
Please remove the use of `com.google.common.collect.Sets` and the dependency to this library from the `Manifest`.

Instead you can use:
```
new HashSet<>(Arrays.asList(new String[] { SERIAL_BRIDGE, ZONE_HANDLER }))
```

martinvw commented:
Or a little shorter:

```java
new HashSet<>(Arrays.asList(SERIAL_BRIDGE, ZONE_HANDLER))
```

martinvw commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialZoneHandler.java:
```diff
@@ -0,0 +1,184 @@
+
+    switch (channelUID.getId()) {
+      case CHANNEL_ZONEPOWER:
+        if (command instanceof OnOffType) {
```
The formatting looks of here, please format using the ESH eclipse profile.

martinvw commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java:
```diff
@@ -0,0 +1,75 @@
+    super.dispose();
+  }
+
+  @Override public void handleCommand(ChannelUID channelUID, Command command) {
```
Please format is such that @Override has its own line, ie please format all the files you added with the Eclipse ESH profile

martinvw commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialZoneHandler.java:
```diff
@@ -0,0 +1,184 @@
+  private void scheduleRefresh() {
+    // Every command requires a status request.
+    if (refreshFuture == null || refreshFuture.isCancelled()) {
+      refreshFuture = scheduler.scheduleAtFixedRate(() -> commander().requestStatus(zone), 0, 100,
```
Please use [scheduleWithFixedDelay](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay(java.lang.Runnable,%20long,%20long,%20java.util.concurrent.TimeUnit)) which is preferred over scheduleAtFixedRate, see [Coding Guidelines D.3](http://docs.openhab.org/developers/development/guidelines.html#d-runtime-behavior)

martinvw replied:
It looks often, is that really needed?

martinvw commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java:
```diff
@@ -0,0 +1,89 @@
+      } catch (FileNotFoundException e) {
+        logger.info("Couldn't find serialPort at {}", serialPort, e);
+        return null;
+      } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

holmes commented on addons/binding/org.openhab.binding.russoundserial/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,92 @@
+    <description>Physical Source Number</description>
+    <state min="0" max="12"/>
+  </channel-type>
+
```
The top ones are `advanced=false`, the bottom are `advanced=true`

holmes commented on addons/binding/org.openhab.binding.russoundserial/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,26 @@
+Bundle-Version: 2.2.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-ClassPath: ., 
+ lib/api-1.2.0.jar,
```
added

holmes commented on addons/binding/org.openhab.binding.russoundserial/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,26 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-ClassPath: ., 
+ lib/api-1.2.0.jar,
+ lib/kotlin-osgi-bundle-1.1.3.jar
```
added

holmes commented on addons/binding/org.openhab.binding.russoundserial/README.md:
```diff
@@ -0,0 +1,98 @@
+There is no auto-discovery. 
+
+#### Bridge
+You must specify the serialPort the Bridge is connected to when setting up the Bridge Thing.
```
Sure. I left them the parameter names to give a hint as what you should be setting.

holmes commented on addons/binding/pom.xml:
```diff
@@ -66,6 +66,7 @@
     <module>org.openhab.binding.rfxcom</module>
     <module>org.openhab.binding.rfxcom.test</module>
     <module>org.openhab.binding.russound</module>
+    <module>org.openhab.binding.russoundserial</module>
```
thanks for the tip, added

holmes commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java:
```diff
@@ -0,0 +1,75 @@
+package org.openhab.binding.russoundserial.handler;
```
updated

holmes commented on addons/binding/org.openhab.binding.russoundserial/about.html:
```diff
@@ -0,0 +1,32 @@
+    indicated below, the terms and conditions of the EPL still apply to any source code in the Content
+    and such source code may be obtained at <a href="http://www.openhab.org/">openhab.org</a>.
+</p>
+
```
added

holmes commented on addons/binding/org.openhab.binding.russoundserial/README.md:
```diff
@@ -0,0 +1,98 @@
+
+## Kotlin
+
+Kotlin is a pretty cool language and the core library that this plugin uses is based on it. Included are OSGI'd versions of the kotlin library for java8.
```
changed

holmes commented on addons/binding/org.openhab.binding.russoundserial/README.md:
```diff
@@ -0,0 +1,98 @@
+You must specify the serialPort the Bridge is connected to when setting up the Bridge Thing.
+
+#### Zones
+You must specify the zoneNumber (0-based) this Thing is targeting. This maps directly to the number on the back of the device. 
```
changed

holmes commented on addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Override
+    public void initialize() {
+        super.initialize();
```
At this point we've established we can make a serial connection, so setting it to `ONLINE` is fine here.

holmes commented:
@martinvw @kaikreuzer @triller-telekom 

I've addressed all the feedback. Formatting the java files makes it hard to figure out where I changed things that were issues. But I got 'em all.

triller-telekom commented on addons/binding/org.openhab.binding.russoundserial/README.md:
```diff
@@ -0,0 +1,98 @@
+zoneLoudness | Switch | Whether the loudness setting is on or off
+zoneTurnOnVolume | Dimmer | This is a write only setting! Determines the volume when the zone is turned on
+
+## Full Example
```
You can have a look here:
http://docs.openhab.org/configuration/things.html#defining-things-using-files

or the README files of other bindings also include a full example as well.

triller-telekom commented:
Oh, one more question that came up while I was browsing through the other bindings... is the current russound binding really that different than your serial one? I mean does it have other things/channels? Otherwise it might make sense to merge the serial related stuff from your binding into the current one.

holmes commented:
@triller-telekom I intiially tried working this into the existing Russound protocol, but it uses a completely different protocol to communicate. I started to see if I could abstract it out, but it's deeply intertwined in that code.

This protocol/implementation are far simpler so I decided making a new one was a cleaner approach.

holmes commented:
Last set of changes were to update README w/ .things example and to update the refresh timeout (only called after a command is issued) from 100ms -> 200ms

holmes commented:
@martinvw @kaikreuzer @triller-telekom

I've addressed all the feedback, squashed commits and am hoping to get this merged! Anything else you'd like me to change?

tmrobert8 commented:
Not to jump in on this but I just noticed it.  The existing Russound binding is already been modified to do all of this (see https://community.openhab.org/t/russound-rnet/27403) by craigh.  The binding uses RNET protocol over a variety of serial connections (native and over several tested IP to serial types).  See [Russound RNET thread](https://community.openhab.org/t/russound-rnet/27403)

I hate seeing work put into this and don't want to really see a duplicate binding happening...

holmes commented:
Well crap.

The RNET thing is a bit misleading, it's just serial, but this probably looks like it does the same thing.

tmrobert8 commented:
Actually RNET is the protocol and serial is the transport for it (you can actually do RIO protocol over the serial transport also).  Sorry about not seeing this sooner - but feel free to help out with the effort (I haven't heard from craigh for awhile - so you may want to watch that thread and help out!)

triller-telekom commented:
@holmes: What is the status of this PR? Did you verify that your serial support now exists in the original russound binding? If it is, do you want to close this PR?

I am really sorry that this happened. Its not nice seeing that people put effort into coding something and then there is another one working on the very same thing :(

Maybe in the future we should encourage the developers to submit work in progress (WIP) PRs so people can see what is being worked on at the moment.

holmes commented:
@triller-telekom I haven't checked the other binding - mine is simple and it works. I couldn't find good documentation nor a build for the other one so I haven't tried.

WIP PRs would be a great thing!

I'll probably just move this to a local repo when the other binding gets merged.

LiberatorUSA commented:
Based on the conversation above, where can those of us who are in need of the serial RNET support find the correct binding? @wborn @holmes 

People will likely only have 1 Russound unit needing either RIO or RNET. As such having a separate binding for each seems acceptable to me. They are essentially different products just from the same vendor.

holmes commented:
The binding isn't published anywhere. Looks like it's out of date w/ the latest code anyways.

I'm finishing up a remodel so I don't have my Russound connected right now. Once I get speakers installed I'll take a look at getting this updated.

J-N-K commented:
@holmes any progress here?

sonydevices commented:
@holmes 
If you are still interested - you can take over craigs work (I have a copy of it) and finish it off and we could merge...

holmes commented:
I ended up getting an MCA-88 and run it over ethernet, so I don't even use this binding anymore.

And honestly, i don't have much of a desire to follow up on this. I can maybe find some time to rebase it on master to get it compiling, but I don't have a way to test it.

Gotta strike while the iron is hot. And that was a long time ago.

J-N-K commented:
@tmrobert8 fondling take over? Otherwise I would close here. 

@holmes Just rebasing doesn‘t help much, sorry. Thanks for your work anyway.

sonydevices commented:
@J-N-K if I can ever get done with the sony binding, I'll swing back around to this one and implement the serial connection.  Craig had a working version that just needs to be merged with the latest and cleaned up for a merge request.  If anyone wants to volunteer - I can give you the code if you want it done much sooner.

@holmes I though the 88 was locked down now?  Are you a vendor or is that not correct?

holmes commented:
We can go ahead and close if no one wants to take over.

And yes, they're locked. But I bought a used one that was unlocked.

J-N-K commented:
Closing as PR is abandoned by original author and noone stepped up to take over.
